### ISSUE TITLE
Massive Latency Increase when Modifying + Reinjecting Packets

### ISSUE THREAD

**Issue by VirtualSmile** [basil00/Divert#159]
Hello again!

After finally getting free time to mess around with WinDivert again, I am running into some troubles (that I'm not sure are normal or not). Mainly I am getting a massive increase in latency when trying to redirect TCP Pings sent via. PSTools. I am using WinDivertSharp by TechnikEmpire found [here](https://github.com/TechnikEmpire/WinDivertSharp).

Here is the code I'm using: 
```
public unsafe void startDivert()
        {
            WinDivertAddress address = new WinDivertAddress();
            WinDivertBuffer packet = new WinDivertBuffer();
            WinDivertParseResult result = null;
            uint packetLength = 0;
            ushort port = 0;

            // Open WinDivert Handle
            IntPtr handle = WinDivert.WinDivertOpen("ip.DstAddr == 210.68.144.12 or ip.SrcAddr == 137.116.149.196", WinDivertLayer.Network, 0, 0);

            while (true)
            {
                switch (address.Direction)
                {
                    case WinDivertDirection.Outbound: 

                        Console.WriteLine("Outbound Packet\n");

                        if (!WinDivert.WinDivertRecv(handle, packet, ref address, ref packetLength))
                        {
                            Console.WriteLine("Failed to Read Outbound Packet! Error Code: " + "{0}\n", Marshal.GetLastWin32Error());
                            continue;
                        }

                        result = WinDivert.WinDivertHelperParsePacket(packet, packetLength);

                        result.IPv4Header->DstAddr = IPAddress.Parse("137.116.149.196");
                        
                        if (result.IPv4Header->DstAddr == IPAddress.Parse("210.68.144.12"))
                        {
                            port = result.TcpHeader->SrcPort;
                            result.TcpHeader->DstPort = (ushort)IPAddress.NetworkToHostOrder((short)17094);
                        }
                        
                        WinDivert.WinDivertHelperCalcChecksums(packet, packetLength, WinDivertChecksumHelperParam.All);

                        if (!WinDivert.WinDivertSend(handle, packet, packetLength, ref address))
                        {
                            Console.WriteLine("Failed to Send Outbound Packet! Error Code: " + "{0}\n", Marshal.GetLastWin32Error());
                            continue;
                        }

                        break;

                    case WinDivertDirection.Inbound:

                        Console.WriteLine("Inbound Packet\n");

                        if (!WinDivert.WinDivertRecv(handle, packet, ref address, ref packetLength))
                        {
                            Console.WriteLine("Failed to Read Inbound Packet! Error Code: " + "{0}\n", Marshal.GetLastWin32Error());
                            continue;
                        }

                        result = WinDivert.WinDivertHelperParsePacket(packet, packetLength);

                        if (result.TcpHeader->DstPort == (ushort)IPAddress.NetworkToHostOrder((short)17094))
                        {
                            result.TcpHeader->DstPort = port;
                        }

                        WinDivert.WinDivertHelperCalcChecksums(packet, packetLength, WinDivertChecksumHelperParam.All);

                        if (!WinDivert.WinDivertSend(handle, packet, packetLength, ref address))
                        {
                            Console.WriteLine("Failed to Send Inbound Packet! Error Code: " + "{0}\n", Marshal.GetLastWin32Error());
                            continue;
                        }

                        break;
                } 
            }
        }
```

The goal is the following:
1. Send a TCP Ping to 210.68.144.12:12100,
2. Parse the packet using WinDivert,
3. Alter the destination address of the ping to 137.116.149.196,
4. Alter the TCP port of the ping to 17094,
5. Re-inject the packet,
6. Handle the return.

I believe that it works, as the TCP pings are effected - but in a very bad way. The latency response should be between 98-100ms, however it is sitting at averages around 3100-3300ms, with it timing out a lot of the time. This only happens if I modify and re-inject the packet, as if I run the passthru example (but in c#) latency is not effected and bandwidth seems okay too. 

I'm starting `startDivert()` on a separate thread (single) using `Task.Run(() => startDivert());`
I'm not sure if it is a threading problem as I did the same for the passthru example and bandwidth and latency did not seem to increase.

Here is an image of the result/issue I am having: 
![Image](https://i.imgur.com/vJRJvvV.png)

It's strange because the returning latency of 180-190ms is actually my latency if I was to TCP ping 210.68.144.12, so it might be WinDivert timing out/re-injecting the packet before it gets modified?

Thank you for all the help again! >~<

EDIT: Can confirm that it only happens when I edit/modify the packet. Even if I parse, get values, print values, re-calc checksums (force to check all) and the re-inject, it doesn't increase latency at all.

**Comment by TechnikEmpire** [comment]
Afaik @basil doesn't know/use c# so won't know what you mean when you're talking about Task.Start. Anyway on that note, Task.Start is not a guarantee of a new thread. It can easily be run synchronously. Only way to be sure is use an actual thread object.

I'm not saying that's the issue, just an observation. Also @basil can cut through a lot of unknown with this info when helping you. The c# library uses raw pointer access so it's just like C in this sense, no language stuff is in the way here. Also even managed objects like packet byte arrays are blitted across the managed/unmanaged boundary so again, it's just like C here as it's a raw pointer with no extra work under the hood.

Basically I'm saying we can probably ignore all info about C# (unless I've a bug in my code) in such questions.

**Comment by VirtualSmile** [comment]
@TechnikEmpire Ah, okay that's good that you cleared things up. I actually started this in C/C++ but I was running into a lot of troubles (as a noob) so I ended up switching to C#. Yep! I was reading up a lot about multi-threading in C# and comparisons between spawning an actual thread vs. a Task and I saw that Task was just better/safer in the majority of scenarios. To be honest I was looking for the C#/.NET alternative to C's "CreateThread" line that I see used in many of the WinDivert samples. I used it simply to start WinDivert not on the UI thread and it did the job so it's not too much of a problem (for now)!

That's good to know! If it makes it any easier for @basil00 the better. I don't think there are any bugs at all.. so we're definitely good on that end haha. Thanks for clearing things up again! (:

**Comment by TechnikEmpire** [comment]
Honestly, the Task/async await system in .NET still confuses me in some cases. I prefer managing my threads myself, not delegating to some black box to schedule things how it sees fit. Generally though when using the task system you need to do "async all the way down" and make sure you're configuring your completion correctly, or else you'll get a traffic jam of synchronous code you're expecting to run parallel.

Just create new Thread() and put your code in a delegate to the constructor just like [here](https://github.com/TechnikEmpire/CitadelCore.Windows/blob/master/CitadelCore.Windows/Diversion/WindowsDiverter.cs#L253) if you're looking for a C/C++ equivalent.

**Comment by VirtualSmile** [comment]
Thanks for the example Technik! I attempted to implement it into my passthru example code and it works great! Although if I set it to use all threads I have in my system it hits 100% (with bandwidth testing) of course haha. 

Unfortunately though when I tried to implement it into the above code for modifying/reinjecting packets, the ping's time out completely :/ I think that might be because the threads are either messing up the order of the packets, or looping it through each thread (along with the fact that 1 thread is already taking 3000+ms) and timing it out.

**Comment by basil00** [comment]
If something is not working, and it cannot be replicated with the standard samples, then it is most likely a bug in your code.  For example, this jumped out at me:

     ...
     result.IPv4Header->DstAddr = IPAddress.Parse("137.116.149.196");
     if (result.IPv4Header->DstAddr == IPAddress.Parse("210.68.144.12"))
     ...

**Comment by TechnikEmpire** [comment]
To answer the question about Last().Start(), it's functionality equivalent to [i], I just felt like using linq there.

**Comment by VirtualSmile** [comment]
@basil00 Thanks for the reply! I was actually thinking about that when I was messing around with the passthru code in C (to try and fix my previous issues I was having there). There was more that I thought I could clean up too so I'll try that tonight and report back on the results. 

@TechnikEmpire I see! I thought it would only start the last result? I tried using it just then and it seemed to only start one thread rather than all threads.. unless I'm missing something here, I'll look back at your example and check again. I think I might've misplaced it's positioning, as I put it outside the for loop so it's only spawning one AFTER every thread has been added to the List.

Thank you both for all the help ^^.

**Comment by VirtualSmile** [comment]
Yep I was right! I managed to solve the issue. The original problem was indeed caused by a bug in the code - the high latency was because the packet wasn't actually being received properly and when a random packet squeezes it's way through it gave me a result. 

The bug/bad code, on top of what you already mentioned above, was this line: 
`result.TcpHeader->DstPort = port;`
I thought that I had to change the destination port to the outbound packet's original source port. Instead, I simply had to change the source port of the INBOUND packet to be the original destination port. I.e:
```
Outbound:
port = result.TcpHeader->DstPort;
result.TcpHeader->DstPort = (ushort)IPAddress.NetworkToHostOrder((short)12100);

Inbound:
result.Tcpheader->SrcPort = port;
```

Thank you both  @basil00 and @TechnikEmpire for all the help again! ^^.

**Comment by TechnikEmpire** [comment]
If you're using threads you need to create local structures for each thread. Do not share objects across threads. Look at my project CitadelCore.Windows, specifically WinDiverter.cs. Trust me there's nothing wrong with WinDivert like this.

**Comment by VirtualSmile** [comment]
@TechnikEmpire Thank you for the help! I managed to do what you suggested with a bit of research and it's working great now <3 Unfortunately the threading didn't fix the problem that I am having overall (which is unstable latency when forwarding game packets using WinDivert), but that was happening even with the socksifier DLL, and was the reason that I attempted to use WinDivert in the first place. Now it's time to work on creating a TCP forwarder/mapper and try to divert to it and hope it remedies the issue!

Thank you again to the both of you!

**Comment by TechnikEmpire** [comment]
I would strongly suggest that you run the performance profiler in visual studio. Enable external code in settings (disable Just My Code). This will show you precisely where CPU cycles are being spent, and I can guarantee you that it's a bottleneck in your logic rather than WinDivert.

Go debug->performance profiler and check CPU cycles and run your program. Make sure you have it set to Release

**Comment by TechnikEmpire** [comment]
Make sure you're using the overlapped API just like I am citadelcore.windows or all those threads will do nothing. They will synchronize at the WinDivertRecv function. You need to use RecEx

**Comment by VirtualSmile** [comment]
Hmm I'll give it a shot! I was looking at trying out SendEx/RecvEx but I didn't know what I/O really meant in this situation, especially considering you had to set a timeout for the packet to be received. I'll take a look using the performance profiler! Thank you for that tip. I was just worried since it's happening to both the DLL and WinDivert that it's something beyond my control. To be honest, it's a weird problem because the situation is like the following:

In the game (Blade and Soul) if I stand still and attack a lot (sends lots of packets) the ping is fine, stable and low. If I was to start moving around while attacking (even very slowly, ends up being less packets in comparison) the ping rises by like 20ms every second until it hits numbers near 3x my expected latency in-game and the lag is "real" i.e you can feel the delay in the game. To remedy this issue before when we were using a DLL, we had to use a local socks proxy that then forwards/connects to a socks server in another location. If we didn't have a local socks server forwarding and instead directly connected to the socks server in another location, the above problem would occur. We tried asking this problem on multiple forums no one could find the reason haha.

Anyway I'll give SendEx and RecvEx a shot as well as the performance profiler! Thanks again Technik ^^.

**Comment by TechnikEmpire** [comment]
Definitely have to use ex functions. You can copy that WinDivert code out of my program, it's fine it's like standard usage of WinDivert so that little portion isn't "mine" in a copyright sense.

You're stuffing all packets into a single processing threats definitely going to have an impact and in fact it'll be worse than using one thread because you're making all threats you spawn sleep and then wake infinitely and a stupid amount of times per second.

**Comment by VirtualSmile** [comment]
Holy crap I see. Well then I really hope that I can get it going and that it at least reduces the effects of the problem. Thank you, haha. Copy and paste is nice but I really need to learn about these things and I'm glad that there's a lot of commenting in your code! Helps a lot.

**Comment by VirtualSmile** [comment]
Alright I managed to (at least I think so) get RecvEx and SendEx going. However, the problem still persists.

I'm not quite sure at what to look for using the performance profiler - when I select my program, I can only check the CPU usage. When I do, the usage is so insanely low. When I stand still and attack (problem doesn't occur, ping is low and smooth) the CPU usage is the exact same as when I do move and attack and the problem arises.

I've attached some images of the results and the created detail report.. if this is better placed elsewhere I would be happy to move it as I don't want to keep beeping on basil. I only fully logged into the game ~ after the 1:40minute mark. So anything past that is the combat/problem testing.
![Image](https://i.imgur.com/7xYKFgH.png)

Results with External Code shown: 
![image](https://i.imgur.com/uJESlOd.png)

Thank you again Technik!

[windivert-perf-profile.zip](https://github.com/basil00/Divert/files/2498088/windivert-perf-profile.zip)

**Comment by TechnikEmpire** [comment]
As basil say, you've got `Console.WriteLine(...)` eating a ton  of CPU cycles. Look at the self CPU time on it. It's eating 99% of your CPU time. Cut all that out. Also I see string.format in there eating a lot of time.

The console is synchronized in .NET AFAIK. Again, this will not only take time writing as basil points out, but will thread lock. Avoid strings as much as possible unless you really understand them. Strings are immutable objects in memory, which means that every time you're changing anything on a string, you're re-allocating the entire string size and modifying that new memory.

Drop everything to do with Console.WriteLine and get away from string.format. You're in packet land, you don't need to be creating and formatting strings in that pipeline in the vast majority of scenarios.

**Comment by TechnikEmpire** [comment]
Calling tostring on every IP address is brutal that's your cause. Use .Equals () method.

https://stackoverflow.com/a/2722628

**Comment by TechnikEmpire** [comment]
Also yes define variables one time. Use static read-only if they never change. That's because every time you do that not only are you parsing but you're reallocating memory every time and you're going to make the garbage collector work overtime. Imagine how many packets per second you're parsing then multiply that by 4 bytes minimum for every version 4 ip address plus the other properties of ipaddress object. In the profiler you should be able to observe memory management and every yellow pin in a garbage collection. I bet it's almost a solid line of yellow pins.

**Comment by TechnikEmpire** [comment]
Since we are entirely in c# land, if you're still stuck open a ticket on WinDivertSharp so we are not spamming basil and everyone else watching.
